### PR TITLE
Fix double-shifted training loss in GitForCausalLM

### LINKED ISSUE #47394: GitForCausalLM double-shifts the training loss and leaks targets across batch rows (regression from #35875)
### System Info

- `transformers`: main (`7ea2320c76`), also present in every release since v4.49
- Code-level bug, weight-independent — reproducible on CPU, no GPU or pretrained weights needed

### Who can help?

@zucchini-nlp

### Description

`GitForCausalLM.forward` shifts the labels manually and then hands the result to `self.loss_function`, which shifts them a **second** time. The model is trained to predict token `t + 2` from position `t` instead of `t + 1`.

`src/transformers/models/git/modeling_git.py:1081-1092`:

```python
if labels is not None:
    # we are doing next-token prediction; shift prediction scores and input ids by one
    num_image_tokens = self.git.encoder.layer[0].attention.self.image_patch_tokens
    shifted_logits = logits[:, num_image_tokens:-1, :].contiguous()
    labels = labels[:, 1:].contiguous()              # shift #1
    loss = self.loss_function(
        shifted_logits.view(-1, self.config.vocab_size),
        labels.view(-1),                             # binds positionally to `labels`
        vocab_size=self.config.vocab_size,
        **kwargs,
    )
```

`GitForCausalLM` resolves `loss_type` to `"ForCausalLM"`, so `self.loss_function` is `ForCausalLMLoss`. Its signature is `(logits, labels, vocab_size, num_items_in_batch, ignore_index, shift_labels)`, so the already-shifted tensor binds to `labels` and `shift_labels` stays `None`. `src/transformers/loss/loss_utils.py:61-64` then does:

```python
if shift_labels is None:
    labels = nn.functional.pad(labels, (0, 1), value=ignore_index)
    shift_labels = labels[..., 1:].contiguous()      # shift #2
```

Two things follow from this. First, the targets are off by one extra position. Second — and this is the part that does not show up in the loss value alone — the manual shift flattens the labels to 1-D *before* the call, so the internal pad-and-slice operates on the flattened tensor and the shape stays consistent (`N → N+1 → N`). No error is raised, and each batch row's final target is silently pulled in from the **next row** of the batch.

### Regression

This was correct until #35875 ("Fix model kwargs", `28f73bc307`), which mechanically swapped the loss implementation but left the manual shift in place:

```diff
-            loss_fct = CrossEntropyLoss()
-            loss = loss_fct(shifted_logits.view(-1, self.config.vocab_size), labels.view(-1))
+            loss = self.loss_function(
+                shifted_logits.view(-1, self.config.vocab_size),
+                labels.view(-1),
```

`CrossEntropyLoss` does not shift, `ForCausalLMLoss` does. This is the same regression mechanism that was fixed for Moonshine in #46784.

I checked the rest of what #35875 touched. On current main only three call sites still combine a manual shift with `self.loss_function`: `csm` is correct (it passes `labels=None, shift_labels=shift_labels`), `moshi` has the same bug but is already covered by the open #36928, and `git` is the remaining unfixed one.

### Reproduction

CPU, no pretrained weights:

```python
import torch
from torch.nn import CrossEntropyLoss

from transformers import GitConfig, GitForCausalLM
from transformers.models.git.configuration_git import GitVisionConfig

torch.manual_seed(0)
vision_config = GitVisionConfig(
    hidden_size=32, num_hidden_layers=2, num_attention_heads=4,
    intermediate_size=37, image_size=32, patch_size=16,
)
config = GitConfig(
    vision_config=vision_config.to_dict(), vocab_size=99, hidden_size=32,
    num_hidden_layers=2, num_attention_heads=4, intermediate_size=37,
)
model = GitForCausalLM(config).eval()

input_ids = torch.randint(3, config.vocab_size, (2, 7))
pixel_values = torch.randn(2, 3, 32, 32)
labels = input_ids.clone()

with torch.no_grad():
    out = model(input_ids=input_ids, pixel_values=pixel_values, labels=labels)

num_image_tokens = model.git.encoder.layer[0].attention.self.image_patch_tokens
shifted_logits = out.logits[:, num_image_tokens:-1, :]
V = config.vocab_size

aligned = CrossEntropyLoss()(shifted_logits.reshape(-1, V), labels[:, 1:].reshape(-1))

# what the code actually computes: a second shift, applied to the flattened targets
flat = labels[:, 1:].reshape(-1)
leaked = torch.nn.functional.pad(flat, (0, 1), value=-100)[1:]
double_shifted = CrossEntropyLoss()(shifted_logits.reshape(-1, V), leaked)

print("model loss            :", out.loss.item())
print("aligned CE (expected) :", aligned.item())
print("double-shifted CE     :", double_shifted.item())
print("matches aligned       :", torch.allclose(out.loss, aligned))
print("matches double-shift  :", torch.allclose(out.loss, double_shifted))
print("row 0 correct targets :", labels[0, 1:].tolist())
print("row 0 actual targets  :", leaked[:6].tolist())
print("row 1 labels          :", labels[1, 1:].tolist())
```

Output on main:

```
model loss            : 4.584834575653076
aligned CE (expected) : 4.646134853363037
double-shifted CE     : 4.584834575653076
matches aligned       : False
matches double-shift  : True
row 0 correct targets : [28, 71, 8, 58, 61, 15]
row 0 actual targets  : [71, 8, 58, 61, 15, 19]
row 1 labels          : [19, 42, 30, 58, 94, 67]
```

The loss matches the double-shifted computation exactly, and row 0's last target (`19`) is row 1's first label.

### Expected behavior

The target for position `t` should be `labels[t + 1]`, and targets must not cross batch-row boundaries.

### Suggested fix

Unlike Moonshine, GIT cannot simply drop the manual shift: `logits` contains `num_image_tokens` leading image positions that have to be sliced off regardless. Passing `shift_labels` explicitly suppresses the internal shift, which is the convention already used by `csm` (`modeling_csm.py:619-621`):

```diff
             shifted_logits = logits[:, num_image_tokens:-1, :].contiguous()
-            labels = labels[:, 1:].contiguous()
+            shift_labels = labels[:, 1:].contiguous()
             loss = self.loss_function(
-                shifted_logits.view(-1, self.config.vocab_size),
-                labels.view(-1),
+                logits=shifted_logits,
+                labels=None,
                 vocab_size=self.config.vocab_size,
+                shift_labels=shift_labels,
                 **kwargs,
             )
```

Keeping the tensors 2-D also removes the cross-row leak, since `ForCausalLMLoss` flattens them itself. With this applied, the repro above prints `4.646135` for both the model loss and the aligned CE. I have the fix and a regression test ready (the full `tests/models/git` suite passes: 190 passed, 297 skipped) and am happy to open a PR if you agree with the direction.

> Disclosure: this investigation and the proposed patch were prepared with AI assistance, and reviewed and verified by me.

### PR DESCRIPTION
**Pull request by Yigtwxx**
[CI](https://transformers-ci.lor-e.huggingface.cool/d/pytest-observability-by-pr/pytest-observability-branch?var-pr=47395)

# What does this PR do?

`GitForCausalLM.forward` shifts the labels manually and then passes them positionally to `self.loss_function`, so they bind to `labels` and `shift_labels` stays `None`. `ForCausalLMLoss` (`loss_utils.py:61-64`) then shifts a second time, and position `t` is trained to predict token `t + 2`.

Because the manual shift flattens the labels to 1-D before the call, the internal pad-and-slice keeps the shape consistent (`N → N+1 → N`), so nothing raises — and each batch row's final target is silently pulled in from the next row.

This regressed in #35875, which swapped `CrossEntropyLoss` (no shift) for `self.loss_function` (shifts) while leaving the manual shift in place. First released in v4.49.0. Same mechanism as the Moonshine fix in #46784.

Unlike Moonshine, GIT cannot just drop the manual shift: `logits` carries `num_image_tokens` leading image positions that must be sliced off regardless. So this PR passes `shift_labels` explicitly to suppress the internal shift, following the convention already used by `csm` (`modeling_csm.py:619-621`), and keeps the tensors 2-D since `ForCausalLMLoss` flattens them itself — which also removes the cross-row leak.

Verified with the repro from the issue (CPU, no pretrained weights). Before:

```
model loss            : 4.584834575653076
aligned CE (expected) : 4.646134853363037
matches aligned       : False
matches double-shift  : True
row 0 correct targets : [28, 71, 8, 58, 61, 15]
row 0 actual targets  : [71, 8, 58, 61, 15, 19]
row 1 labels          : [19, 42, 30, 58, 94, 67]
```

After:

```
model loss            : 4.646134853363037
aligned CE (expected) : 4.646134853363037
matches aligned       : True
matches double-shift  : False
```

Also checked that the `num_items_in_batch` path used by `Trainer` under gradient accumulation still matches the mean reduction when all targets are valid.

Per review, no model-level regression test is added here — the common test coming from #46903 is expected to cover this. The change is source-only.

```
$ pytest tests/models/git/
189 passed, 297 skipped, 2208 subtests passed
```

`ruff check` and `ruff format --check` (pinned 0.14.10) are clean. No `# Copied from` blocks reference this code and GIT has no modular file, so nothing needs regenerating.

Fixes #47394

Disclosure: this fix was implemented with the assistance of a code agent, per my proposal in the linked issue, and reviewed and validated by me.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://huggingface.co/docs/transformers/contributing) and the
      [Pull Request](https://huggingface.co/docs/transformers/pr_checks) checks?
- [x] Was this discussed/approved via a Github issue? https://github.com/huggingface/transformers/issues/47394
- [ ] Did you make sure to update the documentation with your changes according to the [guidelines](https://github.com/huggingface/transformers/tree/main/docs)?
- [ ] Did you write any new necessary [tests](https://huggingface.co/docs/transformers/testing)?

## Who can review?

@zucchini-nlp

### CODE REVIEW

File: src/transformers/models/git/modeling_git.py [hunk 1]
```diff
@@ -1083,11 +1083,12 @@ def forward(
             # we are doing next-token prediction; shift prediction scores and input ids by one
             num_image_tokens = self.git.encoder.layer[0].attention.self.image_patch_tokens
             shifted_logits = logits[:, num_image_tokens:-1, :].contiguous()
-            labels = labels[:, 1:].contiguous()
+            shift_labels = labels[:, 1:].contiguous()
             loss = self.loss_function(
-                shifted_logits.view(-1, self.config.vocab_size),
-                labels.view(-1),
+                logits=shifted_logits,
```
Comment (Sanjays2402) [inline src/transformers/models/git/modeling_git.py:1088]: since forward now pre-shifts and passes `shift_labels` with `labels=None`, the `shift_labels[..., 1:]` step inside `ForCausalLMLoss` is skipped, which is the fix. but the manual `shifted_logits[:, :-1]` + `labels[:, 1:]` pattern here predates that helper. worth checking Git-derived models (e.g. GitForSequenceClassification path or any copy) for the same double-shift, or is git the only one hitting it?

Comment (Yigtwxx) [reply src/transformers/models/git/modeling_git.py:1088]: Good question. I checked, and git is the only model hitting this specific bug.

**Git-derived models:** there are none. `modeling_git.py` defines only `GitVisionModel`, `GitModel` and `GitForCausalLM` — there is no `GitForSequenceClassification`, and no `modular_*.py` file imports from `models/git`, so there is no copy to keep in sync.

**The manual pattern elsewhere is safe.** Models that still do `shift_logits = logits[..., :-1, :]` / `shift_labels = labels[..., 1:]` (gpt2, imagegpt, openai, blip_2, clvp, mamba, falcon_mamba, xlstm, gemma3, llama4, qwen2_audio, granite_speech, modernbert_decoder, ...) pass the result to a plain `nn.CrossEntropyLoss` / `fixed_cross_entropy`, not to `self.loss_function`. `ForCausalLMLoss` is never involved, so there is no second shift. Those are pre-`loss_function` leftovers, not bugs.

**One other model does hit it: `moshi`.** `MoshiForCausalLM.forward` (`src/transformers/models/moshi/modeling_moshi.py#L1003-L1014`) pre-shifts, flattens, and then calls `self.loss_function(shift_logits, shift_labels, vocab_size=...)` positionally — `shift_labels=` is never passed as a keyword, so `ForCausalLMLoss` pads and shifts again (`loss/loss_utils.py#L61-L64`). Because the tensors are already flattened, the shapes stay valid and it fails silently, exactly like git did.

I left that out of this PR to keep the diff scoped to git; happy to open a separate PR for moshi (or fold it in here if a maintainer prefers).

**Review by zucchini-nlp — COMMENTED** [review]
I'd prefer to keep a sinlge PR to fix all models and add a proper tests. A contrib was working on it in https://github.com/huggingface/transformers/pull/46903 (will nudge them again) so let's close this PR

Fixing and reviewing the same issue in all models one-by-one takes too much time for maintainers and for contribs

**Comment by Yigtwxx** [comment]
Thanks for the review, and understood on wanting a single consolidated PR — that makes sense from a maintenance standpoint.

One thing worth flagging before this gets folded into #46903: the two cases have different root paths, and I don't think #46903 as currently written would cover GIT.

#46903 targets encoder-decoder models, where `decoder_input_ids` is built via `shift_tokens_right` so the logits are already aligned with `labels`. Its regression test (`test_encoder_decoder_loss_no_double_shift`) is gated on `config.is_encoder_decoder`, so it skips GIT entirely.

GIT is decoder-only. The double shift here comes from `GitForCausalLM.forward` shifting manually — which it has to do anyway, because `logits` carries `num_image_tokens` leading image positions that must be sliced off — and then passing the result positionally, so it binds to `labels` and `ForCausalLMLoss` shifts again. Dropping the manual shift isn't an option the way it was for Moonshine (#46784); the fix is passing `shift_labels` explicitly.

So a genuinely model-agnostic PR would need to cover both the encoder-decoder path and the decoder-only-with-prefix-tokens path, with a test that isn't gated on `is_encoder_decoder`.

Happy to go either way:
- close this and port the GIT case + an ungated test into #46903 (coordinating with @OmkumarSolanki), or
- leave this open if you'd rather keep the decoder-only variant separate.

Just let me know which you prefer and I'll act on it.

**Comment by zucchini-nlp** [comment]
Oh sorry, indeed the issue here is that we passed manually shifted labels as raw `labels`. Seems to be rather a miss when converging into unified `self.loss`  since GIT is special of itself

Can you delete the test, I dont think it is needed at model level. The common test for encoder-decoder models from another PR should be fine imo, and let's merge

**Comment by Yigtwxx** [comment]
Done — the model-level test is removed, so `tests/models/git/test_modeling_git.py` is now identical to `main` and the PR is source-only (4 insertions, 3 deletions in `modeling_git.py`). Updated the description accordingly.

Locally: `pytest tests/models/git/` → 189 passed, 297 skipped, 2208 subtests passed; `ruff check` and `ruff format --check` clean.

Ready to merge from my side. Note that the common test in #46903 is currently gated on `config.is_encoder_decoder`, so it won't exercise GIT as written — worth ungating it there so this path stays covered.

**Review by zucchini-nlp — APPROVED** [review]
